src/west/app/project.py

```diff
@@ -421,14 +419,19 @@ def delay(func, project):
             # as SHAs, unless they are specifically requested, and then
             # ensures they are only computed once.
             try:
-                if (isinstance(project, ManifestProject) and not
-                        args.manifest_path_from_yaml):
+                if isinstance(project, ManifestProject):
                     # Special-case the manifest repository while it's
                     # still showing up in the 'projects' list. Yet
                     # more evidence we should tackle #327.
-                    path = self.config.get('manifest.path')
-                    apath = abspath(os.path.join(self.topdir, path))
-                    ppath = Path(apath).as_posix()
+                    if args.manifest_path_from_yaml:
+                        path = self.manifest.yaml_path
+                        apath = (abspath(os.path.join(self.topdir, path))
+                                 if path else None)
+                        ppath = Path(apath).as_posix() if apath else None
+                    else:
+                        path = self.config.get('manifest.path')
```

Comment:
shouldn't the `project.path` in the `ManifestProject` be identical to the config `manifest.path`, from what I can tell:
https://github.com/zephyrproject-rtos/west/blob/9dbc5cdf13374daa443d4e6b36bc6114d5356fa7/src/west/manifest.py#L1698
https://github.com/zephyrproject-rtos/west/blob/9dbc5cdf13374daa443d4e6b36bc6114d5356fa7/src/west/manifest.py#L1704
https://github.com/zephyrproject-rtos/west/blob/9dbc5cdf13374daa443d4e6b36bc6114d5356fa7/src/west/manifest.py#L1721
https://github.com/zephyrproject-rtos/west/blob/9dbc5cdf13374daa443d4e6b36bc6114d5356fa7/src/west/manifest.py#L1788-L1789

at best, if they are identical, this code is not needed as it simply uses `self.config.get('manifest.path')` instead of `project.path` that is used in line L436 below.
At worst, if `project.path` != `self.config.get('manifest.path')` then we risk that any code doing:
```
from west.manifest import Manifest
manifest = Manifest.from_file()
manifest_project = manifest.get_projects(['manifest'])
```
compared to `west list manifest` will have different path values, which risk leading to confusion.

Reply:
> shouldn't the project.path in the ManifestProject be identical to the config manifest.path

Yes, but we are trying to get rid of ManifestProject (#327), so I am trying to slowly phase out any in-tree users.

Let me explain my long term goals for after this PR. They are:

#. add a new `west.manifest.Manifest.__init__` kwarg: `legacy_manifest_projects: Optional[bool] = None`. If `legacy_manifest_projects` is `True`, then `Manifest.projects` will include the `ManifestProject` at index 0. If it is `False`, then `Manifest.projects` will **not** include the `ManifestProject`. For compatibility, if `legacy_manifest_projects` is None, then issue a warning about changing behavior and then work the same as if it were `False`.
#. manually set `legacy_manifest_projects` to `True` in static methods in the Manifest class (i.e. `west.manifest.Manifest.from_topdir()`, `west.manifest.Manifest.from_file()`, and `west.manifest.Manifest.from_data()`), so users see no differences and get no warnings (since people in practice use these methods instead of calling `Manifest.__init__` directly
#. Add new **module-level** functions to `west.manifest`: `from_topdir()`, `from_file()`, and `from_data()`, which set `legacy_manifest_projects` to `False`. (I'm talking about a new `west.manifest.from_topdir()` function here, **not** `west.manifest.Manifest.from_topdir()`, etc.)
#. Move all in-tree users over to the new module-level functions, including this part of project.py. Add special case helpers that retain backwards compatibility so the manifest repository is still considered in `west list` output and any other places where it's currently needed (details TBD)
#. Deprecate the static methods (`west.manifest.Manifest.from_topdir()`, etc.)
#. Eventually, remove both the static methods and `ManifestProject`. Some details TBD; I haven't thought it all the way through yet as I'm just trying to fix this one bug in this PR.

At that point, `west list` will not see a `ManifestProject`, and it will need to be using `self.manifest.repo_abspath` and other related fields to access path information for the manifest itself as is being done in the current re-work. I'm laying the ground for that transition to be a bit easier later on in this change by making `west list` get things from the right place for long-term use.

But I take your point that calling self.config is a little ugly here, so I'll add a new `Manifest.repo_path` attribute which caches the configuration option.

Reply:
> Yes, but we are trying to get rid of ManifestProject (https://github.com/zephyrproject-rtos/west/issues/327), so I am trying to slowly phase out any in-tree users.

Fair enough, and thanks for the detailed explanation.

Then I assume the right thing to do in the future when requesting all possible Zephyr modules, is to do:
```
manifest = Manifest.from_file()

# Get the path for the manifest repo 
manifest_repo_path = manifest.repo_path

# Get all other projects.
projects = manifest.get_projects()
```
as the manifest repo will not be part of the project list in future.

Is this something you expect to complete for `0.13` ?

Reply:
> Then I assume the right thing to do in the future when requesting all possible Zephyr modules, is to do:

Not quite. Remember that `west.manifest.Manifest.from_file()` will be the deprecated version that works the same way as it does today, with a `ManifestProject`. The new ways to get a manifest without a `ManifestProject` will be `west.manifest.from_file()` or `west.manifest.from_topdir()`.

Some of the details for zephyr modules are still TBD in my mind. I think one way to handle this cleanly would be to introduce a new `Repository` class, like this:

```py
class Repository:
    '''Attributes:

    - path: relative to topdir
    - abspath: absolute, host form
    - posixpath: absolute, POSIX form
    '''

    def git():
        # run git in the repository

class Project(Repository):
    '''A repository that is also a project in a manifest file,
    but has other attributes as well, like 'url', that Repository
    does not have.'''

class Manifest:
    '''New attribute:

    - repositories: List[Repository], includes all projects
      and the manifest repository
    '''

```

Then in zephyr you could do:

```py
import west.manifest

manifest = west.manifest.from_topdir()
modules = [repository for repository in manifest.repositories if is_zephyr_module(repository)]
for module in modules:
    handle_module(module)
```

Then we could probably keep get_projects() around for backwards compatibility, but return an error unless `legacy_manifest_projects` was `True` at init time. I think the list comprehension style is a little cleaner anyway.

Reply:
> Is this something you expect to complete for 0.13 ?

No; I basically only work on my own west PRs when I get a direct request for a specific feature or fix these days, since there's not a lot of review bandwidth